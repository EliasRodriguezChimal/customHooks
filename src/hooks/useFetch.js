import { useEffect, useRef, useState } from "react"

export const useFetch = (url) => {

    const isMounted = useRef(true);
    const [state, setState] = useState({ data: null, loading: true, error: null });

    useEffect(() => {

        return () => {
            isMounted.current = false;
        }
    }, [])


    useEffect(() => {
        //La linea de abajo servira para mostrar el loading cada vez que se cargue un nuevo quote.
        setState({ data: null, loading: true, error: null });


        fetch(url)
            .then(resp => resp.json())
            .then(data => {
                if (isMounted.current) {
                    setState({
                        loading: false,
                        error: null,
                        data
                    });
                }
            })
            .catch(() => {
                setState({
                    data: null,
                    loading: false,
                    error: 'No se pudo cargar la informacion'
                })
            });
    }, [url])

    return state;
}
