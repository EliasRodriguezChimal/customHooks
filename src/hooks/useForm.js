import { useState } from "react";

//Este hook se encarga de controlar y actualizar los cambios en un input de un form
//gracias al useState (para modificar el valor) y la funcion handleInputChange, que sirve
//para matchearlo con el campo correspondiente dentro del formulario.

export const useForm = (initialState = {}) => {
    const [values, setValues] = useState(initialState);

    const reset = () => {
        setValues(initialState);
    }

    const handleInputChange = ({target}) => {
        setValues({
            ...values,
            [target.name]: target.value
        })
    }

    return [values, handleInputChange, reset];
}