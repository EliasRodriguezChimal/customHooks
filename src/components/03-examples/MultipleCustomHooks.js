import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';

import '../02-useEffect/effects.css';

export const MultipleCustomHooks = () => {
    const { counter, increment } = useCounter(1);

    //Cada vez que el URL de abajo cambia, se vuelve a realizar la peticion y se carga un nuevo quote
    //mediante la logica escrita en el useFetch.
    const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
    const { author, quote } = !!data && data[0];

    return (
        <div>
            <h1>Breaking Bad Quotes</h1>
            <hr />

            {
                loading
                    ?
                    (
                        <div className="alert alert-info text-center">
                            Loading...
                        </div>

                    )
                    :
                    (
                        // NOTA: En Bootstrap ya no es valido "text-right" para alinear el texto a la derecha,
                        // ahora se usa "text-end".
                        <>
                            <blockquote className="blockquote text-end">
                                <p className="mb-0"> {quote} </p>
                                <footer className="blockquote-footer"> {author} </footer>
                            </blockquote>

                            <button
                                className="btn btn-primary"
                                onClick={increment}
                            >
                                Next quote
                            </button>
                        </>
                    )
            }

        </div>
    )
}
