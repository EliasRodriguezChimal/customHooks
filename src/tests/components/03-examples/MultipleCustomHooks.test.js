import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
jest.mock('../../../hooks/useFetch');   //This means that we don't care about the functioning of useFetch, 
//we only care about the results of this hook. The mock function with that path means that we're not using
//the useFetch located inside that file, we're using a 'fake' one instead and avoid calls to the original
//file.

describe('Tests over <MultipleCustomHooks />', () => {
    test('should be displayed correctly', () => {
        useFetch.mockReturnValue({  //This just returns the object written as a 'result' from the fake useFetch
            data: null,
            loading: true,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);

        expect(wrapper).toMatchSnapshot();
    });

    test('should display data', () => {
        useFetch.mockReturnValue({
            data: [{
                author: 'Elias',
                quote: 'UwU Chanclitas'
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);

        expect(wrapper.find('.alert').exists()).toBe(false);    //This just means that nothing with the 'alert' classname should
        //exists in the html snapshot
        expect(wrapper.find('.mb-0').text().trim()).toBe('UwU Chanclitas'); //When you use 'find' function with a classname, this should
        //be written as '.classname', but when you are using find() with tags, this should be written as 'footer'
        expect(wrapper.find('footer').text().trim()).toBe('Elias');
    });
})