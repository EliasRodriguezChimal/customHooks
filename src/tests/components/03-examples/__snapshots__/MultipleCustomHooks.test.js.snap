// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests over <MultipleCustomHooks /> should be displayed correctly 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
</div>
`;
